### PR TITLE
Minor changes to `Utilities::unpack`.

### PR DESCRIPTION
**Pull request by marcfehling**

### CODE REVIEW

**Review by masterleinad — APPROVED** [review]

**Comment by marcfehling** [comment]
Hmm, the generalization of the `unpack` function doesn't work. Turns out that if you serialize an array with just one element, boost adds some extra bytes (in my experiment it was 8 extra bytes).

I'll remove that commit so that only the `constexpr` part is left.

**Review by drwells — APPROVED** [review]

File: include/deal.II/base/utilities.h [hunk 1]
```diff
@@ -1334,7 +1334,11 @@ namespace Utilities
     // see if the object is small and copyable via memcpy. if so, use
     // this fast path. otherwise, we have to go through the BOOST
     // serialization machinery
+#ifdef DEAL_II_HAVE_CXX17
+    if constexpr (std::is_trivially_copyable<T>() && sizeof(T) * N < 256)
+#else
     if (std::is_trivially_copyable<T>() && sizeof(T) * N < 256)
+#endif
```
Comment (bangerth) [inline include/deal.II/base/utilities.h:1341]: I don't think this makes a difference. The condition is a compile-time constant, so whether or not it is marked as `constexpr`, the compiler will remove either the `if` or the `else` branch. The only difference would be if you'd get a syntax error in one of the branches, but you clearly don't. So I don't see a need for this.

**Comment by bangerth** [comment]
I don't like this commit. I think it is unnecessary and just makes the code hard to read. Could we revert it?

As for the one-element thing: I can't see the original commit, so don't know. The point is that we call `pack` and `unpack` symmetrically with the same type. So if you decide that you want to store one-element arrays as just an individual object, then that's ok by using an `if (N==1)` as long as you add the same `if (N==1)` in the unpack function.

File: include/deal.II/base/utilities.h (hunk 1 above)
Comment (masterleinad) [reply include/deal.II/base/utilities.h:1341]: These kinds of changes should stop compilers from complaining that we call `memcpy` for non-trivially-copyable objects. On the other hand, it seems that there wasn't a problem in the first place (or did GCC11 complain?).

Comment (marcfehling) [reply include/deal.II/base/utilities.h:1341]: I was just reviewing code for the serialization feature. I noticed that we have this `constexpr` conditional in the other `pack` and `unpack` functions, so I've put it here as well. So this change is not motivated by any problem.

**Comment by marcfehling** [comment]
> I don't like this commit. I think it is unnecessary and just makes the code hard to read. Could we revert it?

The reason for this change was to do exactly the same in all `pack` and `unpack` functions.

The `constexpr` variant of this conditional has been introduced in https://github.com/dealii/dealii/pull/6475 because `gcc-8` complained.

I would be in favor of leaving it in - users would get warnings if they use c-style arrays with non-trivial types otherwise.

---

> As for the one-element thing: I can't see the original commit, so don't know. 

I've made a backup: https://github.com/marcfehling/dealii/commit/ff317838f6bfcad8a67d20c0cd910bb9b0cdf443

> The point is that we call pack and unpack symmetrically with the same type. So if you decide that you want to store one-element arrays as just an individual object, then that's ok by using an if (N==1) as long as you add the same if (N==1) in the unpack function.

Yes, that would work, but that feels like an unnecessary hack just to get rid of a few lines of code. I don't feel strongly about this - we can leave the variant for c-style arrays. I just hoped that I could avoid code deprecation with this small change, but I would also like to avoid making the pack function more complicated because of this.

**Comment by bangerth** [comment]
No worries, let's leave it as is then!

**Pull request merged**